### PR TITLE
Update the invite page to include a link to the Slack workspace

### PR DESCRIPTION
**Pull request by accorvin**
This addresses #35 

I've additionally updated the development and contributing guides based on some issues I found.

### CODE REVIEW

**Comment by accorvin** [comment]
@dblock in its current state this breaks 3 of the spec tests. I'm hoping you can help me get them working. The tests don't find the field named `name` to fill in the data for, which is a little baffling to me.

File: public/invite.html.erb
Comment (dblock) [inline public/invite.html.erb]: To be super nitpicky, let's call this `team_workspace_url` to be consistent with the other field names.

Add a period since it's a full sentence.

File: public/invite.html.erb
Comment (dblock) [inline public/invite.html.erb]: Since we use JQuery above, let's do `$('#team_link').attr("href", newUrl)`

File: public/js/common.js
Comment (dblock) [inline public/js/common.js]: Move this into the `else if` above just like for `code_already_used`?

File: slack-invite/models/team.rb
Comment (dblock) [inline slack-invite/models/team.rb]: I would break this into two methods:

```
def team_info
    @team_info ||= slack_client.team_info
end

def workspace_url
   team_info&.team.url
end
```

(See if `to_hash...` is unnecessary and whether the above works).

... but see below - we don't want to have to call the API every time.

**Review by dblock — CHANGES_REQUESTED** [review]
Cool! I have a bunch of rather nitpicky comments. Hang in there with me. 

We need to write a (integration) test for this too. LMK if you need help.

**Comment by dblock** [comment]
> @dblock in its current state this breaks 3 of the spec tests. I'm hoping you can help me get them working. The tests don't find the field named `name` to fill in the data for, which is a little baffling to me.

I did a few things to debug this.

1. I made sure my tests pass on master `git checkout master ; rake`.
2. I made sure tests don't pass on this branch.
3. Reproduced with one test, `rspec ./spec/integration/invite_spec.rb:13`.
4. Noticed that the page loaded `http://127.0.0.1:9293/invite?team_id=T0` is blank. The server fails with a 500.
5. Guessed that the reason is that `<%= team && team.workspace_url %>` is failing. This is because there's no slack key, the specs are now calling `workspace_url` which is making an API call to slack.

This is one way to stub it.

```ruby
  before do
    allow_any_instance_of(Slack::Web::Client).to receive(:team_info).and_return(
      team: {
        url: 'https://example.slack.com'
      }
    )
  end
```

Now, I think we should not be calling the Slack API every time this page is loaded, this is an easy way to get blocked by Slack (one can DOS this page by guessing the URL), this information doesn't change often, if at all, and we don't want to do an external call for latency either. Add a field to `Team` called `workspace_url` and we can populate its value on create, so something like this.

```ruby
field :workspace_url, type: String

def team_info
    @team_info ||= slack_client.team_info
end

before_save :update_team_info!

def update_team_info!
   self.workspace_url ||= team_info&.team&.url   
end
```

Existing teams can be backfilled once manually.

For tests try using `, vcr: { cassette_name: 'slack/team_info' }` where appropriate instead of stubbing like I suggested above. I think we used to get team info before, the team_info.yml is in the source code already that has a response for that API call.

Tests that `Fabricate(:team)` can have a value for `workspace_url` added into the fabricator, `team_fabricator.rb`.

**Comment by accorvin** [comment]
@dblock I've finally had time to get back to this. I've addressed your comments and gotten the existing tests working.

I've managed to add an integration test confirming that the workspace_url is linked on the page. I'm trying to add another integration test for the message that is displayed when a user is already in the workspace, but I'm having trouble getting that one working. I'm hoping you can give me a pointer there.

Right now I've split this up into a few commits just so you can see my recent changes more easily. I plan to squash my commits prior to this being merged.

**Comment by dblock** [comment]
> I've managed to add an integration test confirming that the workspace_url is linked on the page. I'm trying to add another integration test for the message that is displayed when a user is already in the workspace, but I'm having trouble getting that one working. I'm hoping you can give me a pointer there.

I think you've confused some workflows. The user goes to the invite page and types their name and email and clicks submit. Then an invitation is created and either `users_admin_invite` is called, or, if the invitation requires approval an admin needs to click something and then `users_admin_invite` is called upon approval. So is the error `already_in_team` returned from the Slack `users_admin_invite` call? Test that first. If that's the case, in the test you've added you expect `Invitation#approve!` to be called, which only happens on approval, which is not the context of that test. Expect `users_admin_invite` to be called and raise `already_in_team`.

The other thing to notice is that it's 2 scenarios: one where the user self-invites (user gets the error message), and the other where the admin approves and then the admin gets a message. I worry that if we do this it becomes a way to check whether any user is a member of a workspace, which I think is not a good idea. Imagine a Slack group called "criminal gang", I can see if dblock[at]dblock[dot]org is a member by filling this form 🤔 

Either way, I think you should reduce this PR to just the adding a link part, your original goal, and let's get that merged? Then you can take your time trying to make the above work.

File: public/js/common.js
Comment (dblock) [inline public/js/common.js:36]: Remove this from this PR.

File: spec/integration/invite_spec.rb
Comment (dblock) [inline spec/integration/invite_spec.rb:15]: Can we be more specific that the page has something like `If you have previously joined this team, click <a id='team_workspace_url'>here</a> to be redirected.`? Maybe check that the `href` on `team_workspace_url` matches this URL.

**Review by dblock — CHANGES_REQUESTED** [review]

**Comment by accorvin** [comment]
My needs around this have changed and I won't be able to get back to this PR any time soon, so closing it.

**Comment by dblock** [comment]
A version of this was added in https://github.com/vestris/slack-invite/commit/fff5ea00f122025d81c7eb6da0b848abb7cb5bf2.

**Pull request closed without merging**
